pkg/util/pki/nameconstraints.go

```diff
@@ -32,86 +37,532 @@ var (
+		var excluded []byte
+		if nameConstraints.Excluded != nil {
+			excludedIPRanges, err := parseCIDRs(nameConstraints.Excluded.IPRanges)
+			if err != nil {
+				return pkix.Extension{}, err
+			}
+			excluded, err = serialiseConstraints(nameConstraints.Excluded.DNSDomains, excludedIPRanges, nameConstraints.Excluded.EmailAddresses, nameConstraints.Excluded.URIDomains)
+			if err != nil {
+				return pkix.Extension{}, err
+			}
+		}
+
+		var b cryptobyte.Builder
+		b.AddASN1(cryptobyte_asn1.SEQUENCE, func(b *cryptobyte.Builder) {
+			if len(permitted) > 0 {
+				b.AddASN1(cryptobyte_asn1.Tag(0).ContextSpecific().Constructed(), func(b *cryptobyte.Builder) {
+					b.AddBytes(permitted)
+				})
+			}
+
+			if len(excluded) > 0 {
+				b.AddASN1(cryptobyte_asn1.Tag(1).ContextSpecific().Constructed(), func(b *cryptobyte.Builder) {
+					b.AddBytes(excluded)
+				})
+			}
+		})
+
+		ext.Value, err = b.Bytes()
 		if err != nil {
 			return pkix.Extension{}, err
 		}
-		nameConstraintsForMarshalling.ExcludedDNSDomains = nameConstraints.Excluded.DNSDomains
-		nameConstraintsForMarshalling.ExcludedIPRanges = excludedIPRanges
-		nameConstraintsForMarshalling.ExcludedEmailAddresses = nameConstraints.Excluded.EmailAddresses
-		nameConstraintsForMarshalling.ExcludedURIDomains = nameConstraints.Excluded.URIDomains
 	}
-	var err error
-	ext.Value, err = asn1.Marshal(nameConstraintsForMarshalling)
-	return ext, err
+	return ext, nil
+}
+
+func doMarshalNameConstraints(nameConstraints *v1.NameConstraints) bool {
+	return nameConstraints != nil &&
+		(nameConstraints.Permitted != nil &&
+			(len(nameConstraints.Permitted.DNSDomains) > 0 ||
+				len(nameConstraints.Permitted.IPRanges) > 0 ||
+				len(nameConstraints.Permitted.EmailAddresses) > 0 ||
+				len(nameConstraints.Permitted.URIDomains) > 0)) ||
+		(nameConstraints.Excluded != nil &&
+			(len(nameConstraints.Excluded.DNSDomains) > 0 ||
+				len(nameConstraints.Excluded.IPRanges) > 0 ||
+				len(nameConstraints.Excluded.EmailAddresses) > 0 ||
+				len(nameConstraints.Excluded.URIDomains) > 0))
+}
+
+func isIA5String(s string) error {
+	for _, r := range s {
+		// Per RFC5280 "IA5String is limited to the set of ASCII characters"
+		if r > unicode.MaxASCII {
+			return fmt.Errorf("x509: %q cannot be encoded as an IA5String", s)
+		}
+	}
+
+	return nil
 }
 
-func parseCIDRs(cidrs []string) ([]net.IPNet, error) {
-	ipRanges := []net.IPNet{}
+func parseCIDRs(cidrs []string) ([]*net.IPNet, error) {
+	ipRanges := []*net.IPNet{}
 	for _, cidr := range cidrs {
 		_, ipNet, err := net.ParseCIDR(cidr)
 		if err != nil {
 			return nil, err
 		}
-		ipRanges = append(ipRanges, net.IPNet{
+		ipRanges = append(ipRanges, &net.IPNet{
 			IP:   ipNet.IP,
 			Mask: ipNet.Mask,
 		})
 	}
 	return ipRanges, nil
 }
 
-func UnmarshalNameConstraints(value []byte) (NameConstraints, error) {
-	var constraints NameConstraints
-	var rest []byte
+func UnmarshalNameConstraints(e pkix.Extension) (*NameConstraints, error) {
```

Comment:
Could you add a comment here too that explains that this is based on x509.go + maybe a permalink?

Reply:
Done